### ISSUE TITLE
Qingming Festival

### ISSUE THREAD

**Issue by Shaenn** [jsauve/AASharp#11]
Hello,

What amazing repository !

I'm searching a way to calculate the 5th solar term for the chinese calendar.
I know there is some calculation to do in order to get the gregorian date, but I have no idea to do this :/.
I think your libraries can do this, but have you any idea ?

If you can help we with your project it will be so cool ;)

Some link :
https://en.wikipedia.org/wiki/Solar_term
https://en.wikipedia.org/wiki/Qingming_Festival
http://www.math.nus.edu.sg/aslaksen/gem-projects/hm/Chinese_Calendar.pdf

Best regards,

Shaenn

**Comment by jsauve** [comment]
Hi @Shaenn,

The .NET Framework contains a ChineseLunisolarCalendar class:
https://docs.microsoft.com/en-us/dotnet/api/system.globalization.chineselunisolarcalendar

**Comment by jsauve** [comment]
In other words, I believe you are able to use DateTime to convert between GregorianCalendar and EastAsianLunisolarCalendar dates.

**Comment by Shaenn** [comment]
Hi, thanks,

But, there is no way in .NET Framework to get the datetime corresponding to the 5th solar term.
The 5th solar term, is 15° after the spring equinox.

I'm searching a way to calcul a Julian day from the solar longitude and a year and convert this into a datetime.

PS : I' finding this site, but there is no algorithm available (http://wise-obs.tau.ac.il/~eran/Wise/Util/SolLon.html) . Maybe your librairie can do this ? 

Shaenn

**Comment by jsauve** [comment]
Hi @Shaenn,

A few questions for you:

1) In which timezone is the observation made to determine the 15 degrees?

2) What time of day is used for the observation? 00:00? 12:00?

3) What data can I use as the source of truth for the calculations? I looked at the data from the following two sites, but they present conflicting data (the 2021 date in particular):
https://www.travelchinaguide.com/essential/holidays/qingming.htm
https://www.timeanddate.com/holidays/china/qing-ming-jie
In other words, what source of information is considered the de facto for the dates that will be marked on the Chinese calendar as QingMing? I ask this because I need some source of information to which I can compare the results that the algorithm produces.

**Comment by jsauve** [comment]
...the conflicts in the reported QingMing dates may possibly be due to the socially _observed_ holiday date versus the date that the 15 degrees solar ecliptic longitude actually occurs.

**Comment by Shaenn** [comment]
Oh, in my research I havent detected this conflict.
I see in another site (https://www.yourchineseastrology.com/holidays/qingming-festival/#Date%20and%20Holiday%20Durations) in 2021 this is on April the 4th.

3. We can consider https://www.timeanddate.com/holidays/china/qing-ming-jie as reference for calcul.

1. this festival was so old, I did'nt think that they use any timezone, but if you want one try with 'China standard Time' or otherwise try with UTC.

2. In China, it seems to be midnight for the beginning of observation of day.

Shaenn.

**Comment by jsauve** [comment]
@Shaenn, ok, I will try to calculate later this evening (it's 2pm here at the moment). I did already write some code to perform your calculations, but there were some minor inaccuracies. I will try to get it more accurate this evening.

**Comment by Shaenn** [comment]
Oh Nice ! 
In France it's 9pm and it's time for the night ! See u tomorrow!

Shaenn

**Comment by jsauve** [comment]
@Shaenn, I created a simple console program in order to try to work out a solution to your question. The code loops over the range of years and tests the 4th - 6th of April to see if the Sun's ecliptic longitude is at least 15 degrees. The first day encountered with 15 degrees is considered a match and the inner loop breaks, moving on to the next year.
```
using System;
using System.Collections.Generic;
using System.Linq;
using AASharp;

namespace QingMing
{
    class MainClass
    {
        public static void Main(string[] args)
        {
            var successes = 0;

            var yearRange = Enumerable.Range(2000, 50);
            
            Console.WriteLine($"Year | Expect | Calc | Fail");

            foreach (var year in yearRange)
            {
                var calculatedDay = 4;

                for (int day = calculatedDay; day < 7; day++)
                {
                    var zonelessDateTime = DateTime.SpecifyKind(new DateTime(year, 4, day, 0, 0, 0), DateTimeKind.Unspecified);

                    var chinaDateTimeAsUtc = TimeZoneInfo.ConvertTimeToUtc(zonelessDateTime, TimeZoneInfo.FindSystemTimeZoneById("Asia/Shanghai"));

                    var dayDurationStart = chinaDateTimeAsUtc;
                    var dayDurationEnd = dayDurationStart.AddDays(1);

                    AASDate dateEndSunCalc = new AASDate(dayDurationEnd.Year, dayDurationEnd.Month, dayDurationEnd.Day, dayDurationEnd.Hour, dayDurationEnd.Minute, dayDurationEnd.Second, true);

                    double JDEndSun = dateEndSunCalc.Julian + AASDynamicalTime.DeltaT(dateEndSunCalc.Julian) / 86400.0;
                    double SunEndLong = AASSun.ApparentEclipticLongitude(JDEndSun, true);

                    if (SunEndLong >= 15)
                    {
                        calculatedDay = day;
                        break;
                    }
                }

                string x = (TestValues[year] != calculatedDay) ? "X" : "";

                Console.WriteLine($"{year} |      {TestValues[year]} | {calculatedDay}    | {x}");

                //Console.WriteLine($"Apr {calculatedDay}, {year}");
                if (TestValues[year] == calculatedDay)
                    successes++;
            }
            Console.WriteLine();
            Console.WriteLine($"Success rate: {(double)successes/TestValues.Count * 100}%");
        }

        // 2000-2049 values from https://www.timeanddate.com/holidays/china/qing-ming-jie
        public static Dictionary<int, int> TestValues = new Dictionary<int, int>()
        {   
            {2000,4},
            {2001,5},
            {2002,5},
            {2003,5},
            {2004,4},
            {2005,5},
            {2006,5},
            {2007,5},
            {2008,4},
            {2009,5},
            {2010,5},
            {2011,5},
            {2012,4},
            {2013,4},
            {2014,5},
            {2015,5},
            {2016,4},
            {2017,4},
            {2018,5},
            {2019,5},
            {2020,4},
            {2021,5},
            {2022,5},
            {2023,5},
            {2024,4},
            {2025,5},
            {2026,5},
            {2027,5},
            {2028,4},
            {2029,5},
            {2030,5},
            {2031,5},
            {2032,4},
            {2033,5},
            {2034,5},
            {2035,5},
            {2036,4},
            {2037,5},
            {2038,5},
            {2039,5},
            {2040,4},
            {2041,5},
            {2042,5},
            {2043,5},
            {2044,4},
            {2045,5},
            {2046,5},
            {2047,5},
            {2048,4},
            {2049,5},
        };
    }
}
```
The calculated results seem to only match 78% of the data for the 2000-2049 range found here: 
https://www.timeanddate.com/holidays/china/qing-ming-jie

```
Year | Expect | Calc | Fail
2000 |      4 | 4    | 
2001 |      5 | 5    | 
2002 |      5 | 5    | 
2003 |      5 | 5    | 
2004 |      4 | 4    | 
2005 |      5 | 5    | 
2006 |      5 | 5    | 
2007 |      5 | 5    | 
2008 |      4 | 4    | 
2009 |      5 | 4    | X
2010 |      5 | 5    | 
2011 |      5 | 5    | 
2012 |      4 | 4    | 
2013 |      4 | 4    | 
2014 |      5 | 5    | 
2015 |      5 | 5    | 
2016 |      4 | 4    | 
2017 |      4 | 4    | 
2018 |      5 | 5    | 
2019 |      5 | 5    | 
2020 |      4 | 4    | 
2021 |      5 | 4    | X
2022 |      5 | 5    | 
2023 |      5 | 5    | 
2024 |      4 | 4    | 
2025 |      5 | 4    | X
2026 |      5 | 5    | 
2027 |      5 | 5    | 
2028 |      4 | 4    | 
2029 |      5 | 4    | X
2030 |      5 | 5    | 
2031 |      5 | 5    | 
2032 |      4 | 4    | 
2033 |      5 | 4    | X
2034 |      5 | 5    | 
2035 |      5 | 5    | 
2036 |      4 | 4    | 
2037 |      5 | 4    | X
2038 |      5 | 5    | 
2039 |      5 | 5    | 
2040 |      4 | 4    | 
2041 |      5 | 4    | X
2042 |      5 | 4    | X
2043 |      5 | 5    | 
2044 |      4 | 4    | 
2045 |      5 | 4    | X
2046 |      5 | 4    | X
2047 |      5 | 5    | 
2048 |      4 | 4    | 
2049 |      5 | 4    | X

Success rate: 78%
```
The year 2021, for example, does not match the value found at timeanddate.com, but it _does_ match the value found at https://www.yourchineseastrology.com/holidays/qingming-festival/#Date%20and%20Holiday%20Durations.

So, I don't which data is correct.

**Comment by Shaenn** [comment]
@jsauve 78% of success is already better than nothing actually !
Thanks a lot,

I will continue my work.

Best regards, 

Shaenn

**Comment by Shaenn** [comment]
Hello,

I try some modification, but I don't know what I'm doing ! :D

I've replaced this : AASSun.ApparentEclipticLongitude( JDEndSun, true );
by AASSun.GeometricEclipticLongitudeJ2000( JDEndSun, true );
And I've replaced : new DateTime( year, 4, day, 0, 0, 0 )
by new DateTime( year, 4, day, 4, 0, 0 )

And I Get 94%.

I repeat, I don't know what I'm doing, maybe help you in research on the 100% if it's possible. :s

Shaenn

**Comment by elendil-software** [comment]
@jsauve Should we include that in AA# ?

**Comment by jsauve** [comment]
@elendil-software, Do you mean include calculations for the 24 solar terms? (i.e., the moment/date at which the solar ecliptic longitude is at intervals of 15 degrees? Or do you specifically mean calculating QingMing?

If you mean calculating for QingMing, I think we first need to know the authoritative formula for calculating for it. Are solar terms specific to the observer? Are they calculated from GMT or local timezones? Is the interval determined after 00:00, or after 12:00, or after the solar azimuth of that day? I'm just not sure.

According to the QingMing dates provided by @Shaenn (https://www.timeanddate.com/holidays/china/qing-ming-jie), I have not yet been able to calculate QingMing with 100% accuracy. It sounds as though @Shaenn has been able to get it close (94%) by using a timezone offset of +4, and the GeometricEclipticLongitudeJ2000() method.

In general, I'm leery of embedding QingMing calculations directly in AASharp. In my opinion, the library should remain generalized. It can be _used_ to calculate QingMing, but in the interest of remaining fairly agnostic, I believe calculations for specific cultural events should be at a minimum. By that same thinking, the AAEaster class should not exist, but it's included because it's a part of the original AA+ lib.

If anything, it would great to provide functionality that calculates the AASDate for a given solar ecliptic longitude value. Essentially a reverse of the GeometricEclipticLongitude...() methods. However, I'm not sure if that's reasonable or feasible. Even calculating for a specific date day is probably an efficiency of O(n).

What are you thoughts?

**Comment by jsauve** [comment]
@Shaenn, fo your purposes, you could probably just use the algorithm you've worked out to get the 94% success rate. All the non-matching dates are in the past now anyway. Unless those past dates are important to your project. It depends on how you're using the data.

**Comment by jsauve** [comment]
@Shaenn, when I update the test data with additional future years (2000-2099), the results start to become inaccurate again from 2060 onwards. So I advise caution in how you use the calculations.

```
Year | Expect | Calc | Fail
2000 |      4 | 4    | 
2001 |      5 | 4    | X
2002 |      5 | 5    | 
2003 |      5 | 5    | 
2004 |      4 | 4    | 
2005 |      5 | 4    | X
2006 |      5 | 5    | 
2007 |      5 | 5    | 
2008 |      4 | 4    | 
2009 |      5 | 4    | X
2010 |      5 | 5    | 
2011 |      5 | 5    | 
2012 |      4 | 4    | 
2013 |      4 | 4    | 
2014 |      5 | 5    | 
2015 |      5 | 5    | 
2016 |      4 | 4    | 
2017 |      4 | 4    | 
2018 |      5 | 5    | 
2019 |      5 | 5    | 
2020 |      4 | 4    | 
2021 |      5 | 5    | 
2022 |      5 | 5    | 
2023 |      5 | 5    | 
2024 |      4 | 4    | 
2025 |      5 | 5    | 
2026 |      5 | 5    | 
2027 |      5 | 5    | 
2028 |      4 | 4    | 
2029 |      5 | 5    | 
2030 |      5 | 5    | 
2031 |      5 | 5    | 
2032 |      4 | 4    | 
2033 |      5 | 5    | 
2034 |      5 | 5    | 
2035 |      5 | 5    | 
2036 |      4 | 4    | 
2037 |      5 | 5    | 
2038 |      5 | 5    | 
2039 |      5 | 5    | 
2040 |      4 | 4    | 
2041 |      5 | 5    | 
2042 |      5 | 5    | 
2043 |      5 | 5    | 
2044 |      4 | 4    | 
2045 |      5 | 5    | 
2046 |      5 | 5    | 
2047 |      5 | 5    | 
2048 |      4 | 4    | 
2049 |      5 | 5    | 
2050 |      5 | 5    | 
2051 |      5 | 5    | 
2052 |      4 | 4    | 
2053 |      5 | 5    | 
2054 |      5 | 5    | 
2055 |      5 | 5    | 
2056 |      4 | 4    | 
2057 |      5 | 5    | 
2058 |      5 | 5    | 
2059 |      5 | 5    | 
2060 |      4 | 5    | X
2061 |      5 | 5    | 
2062 |      5 | 5    | 
2063 |      5 | 5    | 
2064 |      4 | 5    | X
2065 |      5 | 5    | 
2066 |      5 | 5    | 
2067 |      5 | 5    | 
2068 |      4 | 5    | X
2069 |      5 | 5    | 
2070 |      5 | 5    | 
2071 |      5 | 5    | 
2072 |      4 | 5    | X
2073 |      5 | 5    | 
2074 |      5 | 5    | 
2075 |      5 | 5    | 
2076 |      4 | 5    | X
2077 |      5 | 5    | 
2078 |      5 | 5    | 
2079 |      5 | 5    | 
2080 |      4 | 5    | X
2081 |      5 | 5    | 
2082 |      5 | 5    | 
2083 |      5 | 5    | 
2084 |      4 | 5    | X
2085 |      5 | 5    | 
2086 |      5 | 5    | 
2087 |      5 | 5    | 
2088 |      4 | 5    | X
2089 |      5 | 5    | 
2090 |      5 | 5    | 
2091 |      5 | 5    | 
2092 |      4 | 5    | X
2093 |      5 | 5    | 
2094 |      5 | 5    | 
2095 |      5 | 5    | 
2096 |      4 | 5    | X
2097 |      5 | 5    | 
2098 |      5 | 5    | 
2099 |      5 | 6    | X

Success rate: 86%
```

**Comment by jsauve** [comment]
@Shaenn, this online solar term calculator:
https://keisan.casio.com/exec/system/1227965257
...produces different results than those found here:
https://www.timeanddate.com/holidays/china/qing-ming-jie

For example, try the year 2021 with an offset of +8 (for the chinese timezone). The result from keisan.casio.com is the 4th of April, whereas timeanddate.com says that it's 5th of April. So, there is some discrepancy.

I believe that the AA# does a good job of calculating which day sees the solar ecliptic at 15 degrees. The problem is that perhaps the official date of QingMing is not calculated exactly according to the solar ecliptic. Either that or there is some variable that is currently unknown to us. I'm having difficulty finding an authoritative source.

**Comment by Shaenn** [comment]
@jsauve Ok, 
So with data of 2000-2099, you use your algorithm or the modified version ?
86% for 100 years, it's ok for me.
I'm agree with you QingMing is not calculated exactly according to the solar ecliptic.
There is so much website with no same data. I think Chinese government adjust this date manually, at the start of year maybe.

Shaenn.

**Comment by jsauve** [comment]
@Shaenn, this will produce the 86%:

```
using System;
using System.Collections.Generic;
using System.Linq;
using AASharp;

namespace QingMing
{
    class MainClass
    {
        public static void Main(string[] args)
        {
            var successes = 0;

            var yearRange = Enumerable.Range(2000, 100);

            Console.WriteLine($"Year | Expect | Calc | Fail");
             
            foreach (var year in yearRange)
            {
                var calculatedDay = 4;

                for (int day = calculatedDay; day < 7; day++)
                {
                    var observationDateTimeUtc = DateTime.SpecifyKind(new DateTime(year, 4, day, 0, 0, 0), DateTimeKind.Utc);

                    var endOfDay = observationDateTimeUtc.AddDays(1);

                    AASDate dateEndSunCalc = new AASDate(endOfDay.Year, endOfDay.Month, endOfDay.Day, endOfDay.Hour, endOfDay.Minute, endOfDay.Second, true);

                    double JDEndSun = dateEndSunCalc.Julian + AASDynamicalTime.DeltaT(dateEndSunCalc.Julian) / 86400.0;

                    double SunEndLong = AASSun.GeometricEclipticLongitudeJ2000(JDEndSun, true);

                    if (SunEndLong >= 15)
                    {
                        calculatedDay = day;
                        break;
                    }
                }

                string x = (TestValues[year] != calculatedDay) ? "X" : "";

                Console.WriteLine($"{year} |      {TestValues[year]} | {calculatedDay}    | {x}");

                if (TestValues[year] == calculatedDay)
                    successes++;
            }
            Console.WriteLine();
            Console.WriteLine($"Success rate: {(double)successes/TestValues.Count * 100}%");
        }

        // 2000-2099 values from https://www.timeanddate.com/holidays/china/qing-ming-jie
        public static Dictionary<int, int> TestValues = new Dictionary<int, int>()
        {   
            {2000,4},
            {2001,5},
            {2002,5},
            {2003,5},
            {2004,4},
            {2005,5},
            {2006,5},
            {2007,5},
            {2008,4},
            {2009,5},
            {2010,5},
            {2011,5},
            {2012,4},
            {2013,4},
            {2014,5},
            {2015,5},
            {2016,4},
            {2017,4},
            {2018,5},
            {2019,5},
            {2020,4},
            {2021,5},
            {2022,5},
            {2023,5},
            {2024,4},
            {2025,5},
            {2026,5},
            {2027,5},
            {2028,4},
            {2029,5},
            {2030,5},
            {2031,5},
            {2032,4},
            {2033,5},
            {2034,5},
            {2035,5},
            {2036,4},
            {2037,5},
            {2038,5},
            {2039,5},
            {2040,4},
            {2041,5},
            {2042,5},
            {2043,5},
            {2044,4},
            {2045,5},
            {2046,5},
            {2047,5},
            {2048,4},
            {2049,5},
            {2050,5},
            {2051,5},
            {2052,4},
            {2053,5},
            {2054,5},
            {2055,5},
            {2056,4},
            {2057,5},
            {2058,5},
            {2059,5},
            {2060,4},
            {2061,5},
            {2062,5},
            {2063,5},
            {2064,4},
            {2065,5},
            {2066,5},
            {2067,5},
            {2068,4},
            {2069,5},
            {2070,5},
            {2071,5},
            {2072,4},
            {2073,5},
            {2074,5},
            {2075,5},
            {2076,4},
            {2077,5},
            {2078,5},
            {2079,5},
            {2080,4},
            {2081,5},
            {2082,5},
            {2083,5},
            {2084,4},
            {2085,5},
            {2086,5},
            {2087,5},
            {2088,4},
            {2089,5},
            {2090,5},
            {2091,5},
            {2092,4},
            {2093,5},
            {2094,5},
            {2095,5},
            {2096,4},
            {2097,5},
            {2098,5},
            {2099,5},
        };
    }
}
```

**Comment by jsauve** [comment]
@Shaenn, notice that I'm using GMT without adjusting the timezone for China. This suggests that the QingMing calculation may possibly made from GMT. But again, I've had a hard time finding a definitive answer on that. But the calculations match the timeanddate.com data by 86% when calculated from GMT. Adjusting the timezone offset seems to make the results worse when further from GMT.

**Comment by Shaenn** [comment]
@jsauve Ok ty for your algorithm !

86% it's ok for me !!

Thanks,

Shaenn ;-)

**Comment by elendil-software** [comment]
> 
> 
> @elendil-software, Do you mean include calculations for the 24 solar terms? (i.e., the moment/date at which the solar ecliptic longitude is at intervals of 15 degrees? Or do you specifically mean calculating QingMing?
> 
> If you mean calculating for QingMing, I think we first need to know the authoritative formula for calculating for it. Are solar terms specific to the observer? Are they calculated from GMT or local timezones? Is the interval determined after 00:00, or after 12:00, or after the solar azimuth of that day? I'm just not sure.
> 
> According to the QingMing dates provided by @Shaenn (https://www.timeanddate.com/holidays/china/qing-ming-jie), I have not yet been able to calculate QingMing with 100% accuracy. It sounds as though @Shaenn has been able to get it close (94%) by using a timezone offset of +4, and the GeometricEclipticLongitudeJ2000() method.
> 
> In general, I'm leery of embedding QingMing calculations directly in AASharp. In my opinion, the library should remain generalized. It can be _used_ to calculate QingMing, but in the interest of remaining fairly agnostic, I believe calculations for specific cultural events should be at a minimum. By that same thinking, the AAEaster class should not exist, but it's included because it's a part of the original AA+ lib.
> 
> If anything, it would great to provide functionality that calculates the AASDate for a given solar ecliptic longitude value. Essentially a reverse of the GeometricEclipticLongitude...() methods. However, I'm not sure if that's reasonable or feasible. Even calculating for a specific date day is probably an efficiency of O(n).
> 
> What are you thoughts?

Hi,

I mean calculating QingMing.

But that could be other things. I think this is the problem with AA# and AA+ as well : you can't do anything if you don't know how to compute what you are looking for. You need to know what you are doing. It's not a bad thing to understand how it's work but can be annoying if you just want to compute some information.

I also think it's a good thing to keep AA# code isolated, but having some helper class/methods can't be bad. Maybe in a separated project ?

**Comment by jsauve** [comment]
@elendil-software, I agree, having some helpers or a separate project isn't bad at all. You're right that it would make AA# more accessible to people who are not interested in the details of celestial mechanics, and yet still want to use the library for specific use cases.

I don't presume to have any controlling authority over AA#, even though I'm the owner of the repo. So, I'll defer to your judgment. If it was up to me, a separate project that references the AA# Nuget would be ideal. It would be a clean separation from AA#, and yet still provide useful meaningful calculations to people, such as for QingMing. 

As far as solar term calculation goes, I'd say that if we're going to calculate QingMing, we might as well calculate all 24 solar terms.

**Comment by elendil-software** [comment]
@jsauve Ok. I will create a new repository for that. We just need to find a name. Maybe "AASharp Extensions" ? Or "AASharp Utilities" ?
